packages/block-library/src/table-of-contents/utils.js

```diff
@@ -0,0 +1,135 @@
+/**
+ * WordPress dependencies
+ */
+import { select } from '@wordpress/data';
+import { cleanForSlug } from '@wordpress/editor';
+
+export function linearToNestedList( array, index = 0 ) {
+	const returnValue = [];
+
+	array.forEach( function( heading, key ) {
```

Comment:
I think a `.reduce` might actually make more sense in this case.

Reply:
Yes, even better, as we're building that `returnValue` thing :)

Reply:
But the loop needs to look ahead to the next value (with the `key` parameter) to check if it will return a value. I didn't write this (it's uplifted from the old PR or the original plugin) but the use of `forEach` here seems appropriate to me.